marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Utxo.hs

```diff
@@ -0,0 +1,683 @@
+{-# LANGUAGE AllowAmbiguousTypes   #-}
+{-# LANGUAGE DeriveGeneric         #-}
+{-# LANGUAGE DerivingStrategies    #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE PackageImports        #-}
+{-# LANGUAGE PatternSynonyms       #-}
+{-# LANGUAGE QuasiQuotes           #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE StrictData            #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE UndecidableInstances  #-}
+
+{-# OPTIONS_GHC -fno-warn-redundant-constraints #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Marconi.ChainIndex.Experimental.Indexers.Utxo where
+
+import Control.Lens (folded, imap, makeLenses, view, (^.), (^..))
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import Data.Either (fromRight)
+import Data.Foldable (fold, foldl')
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.Ord ()
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.Text qualified as Text
+import Database.SQLite.Simple (NamedParam ((:=)))
+import Database.SQLite.Simple qualified as SQL
+import Database.SQLite.Simple.FromRow (FromRow (fromRow), field)
+import Database.SQLite.Simple.ToField (ToField (toField))
+import Database.SQLite.Simple.ToRow (ToRow (toRow))
+import Text.RawString.QQ (r)
+
+import Cardano.Api ()
+import Cardano.Api qualified as C
+import Cardano.Api qualified as Core
+import Cardano.Api.Shelley qualified as C
+import GHC.Generics (Generic)
+import Marconi.ChainIndex.Orphans ()
+import Marconi.ChainIndex.Types (SecurityParam (SecurityParam),
+                                 TargetAddresses, TxOut,
+                                 pattern CurrentEra)
+import Marconi.Core.Experiment qualified as Core
+
+data Utxo = Utxo
+  { _utxoAddress          :: !C.AddressAny
+  , _utxoTxIn             :: !C.TxIn
+  , _utxoDatum            :: !(Maybe C.ScriptData)
+  , _utxoDatumHash        :: !(Maybe (C.Hash C.ScriptData))
+  , _utxoValue            :: !C.Value
+  , _utxoInlineScript     :: !(Maybe C.ScriptInAnyLang)
+  , _utxoInlineScriptHash :: !(Maybe C.ScriptHash)
+  } deriving (Show, Eq, Generic)
+
+$(makeLenses ''Utxo)
+
+instance Ord Utxo where
+  compare (Utxo addr txin _ _ _ _ _) (Utxo addr' txin' _ _ _ _ _) =
```

Comment:
Spotted this during the demo: isn't txin enough here as it uniquely identifies a (U)TxO?

Reply:
No as you want to sort by address first: `txIn` is used only if the addresses are equal.

Reply:
Or maybe we don't care about ordering by address first, it was my assumption though.

Reply:
I think txin should be enough, not sure if we ever sort by address. I'll leave it for now to think about a bit more.

Reply:
I guess it's the Set/Map operations where the Ord instance is useful? If yes then I'd go with just the TxIn. It would fit the use case of where you have a Map and you want to do a lookup by TxIn. If you add address to the mix then that AFAIU is superfluous as there are many TxIn's with the same address.